src/prototype.rs

```diff
@@ -1697,6 +1701,125 @@ fn scan_link_title(scanner: &mut InlineScanner) -> Option<String> {
     None
 }
 
+// TODO: cows?
+fn scan_autolink(scanner: &mut InlineScanner) -> Option<String> {
+    let save = scanner.clone();
+    let scans = scan_uri(scanner).or_else(|| {
+        *scanner = save.clone();
+        scan_email(scanner)
+    });
+    if let Some(uri) = scans {
+        if scanner.scan_ch(b'>') {
+            return Some(uri);
+        }
+    }
+    *scanner = save;
+    None
+}
+
+// must return scanner to original state
+// TODO: such invariants should probably be captured by the type system
+fn scan_uri(scanner: &mut InlineScanner) -> Option<String> {
+    // FIXME: don't use strings?
+    let mut uri = String::new();
+
+    // scheme's first byte must be an ascii letter
+    let first = scanner.next()?;
+    if !is_ascii_alpha(first) {
+        return None;
+    } else {
+        uri.push(first as char);
+    }
+
+    while let Some(c) = scanner.next() {
+        match c {
+            c if is_ascii_alphanumeric(c) => uri.push(c as char),
+            c @ b'.' | c @ b'-' | c @ b'+' => uri.push(c as char),
+            b':' => {uri.push(c as char); break; }
+            _ => {
+                return None;
+            }
+        }
+    }
+
+    // scheme length must be between 2 and 32 characters long. scheme
+    // must be followed by colon
+    if uri.len() < 3 || uri.len() > 33  {
+        return None;
+    }
+
+    let mut ended = false;
+    while let Some(c) = scanner.next() {
+        match c {
+            b'\0' ... b' ' => {
+                ended = true;
+            }
+            b'>' | b'<' => break,
+            _ if ended => return None,
+            c => uri.push(c as char),
+        }
+    };
+    scanner.unget();
+
+    Some(uri)
+}
+
+fn scan_email(scanner: &mut InlineScanner) -> Option<String> {
+    // using a regex library would be convenient, but doing it by hand is not too bad
+
+    // FIXME: don't use strings?
```

Comment:
I think `String` is ok here. We can look at it in a performance pass later.

Reply:
Ok